core/domain/exp_services.py

```diff
@@ -1531,40 +1598,64 @@ def regenerate_exploration_and_contributors_summaries(exploration_id):
         exploration_id: str. ID of the exploration.
     """
     exploration = exp_fetchers.get_exploration_by_id(exploration_id)
-    exp_summary = _compute_summary_of_exploration(exploration)
+    exp_rights = exp_models.ExplorationRightsModel.get(
+        exploration_id, strict=False)
+    exp_summary_model = exp_models.ExpSummaryModel.get(
+        exploration_id, strict=False)
+    exp_summary = compute_summary_of_exploration(
+        exploration, exp_rights, exp_summary_model)
     exp_summary.contributors_summary = (
         compute_exploration_contributors_summary(exp_summary.id))
     save_exploration_summary(exp_summary)
 
 
-def _compute_summary_of_exploration(exploration):
+def compute_summary_of_exploration(
+    exploration, exp_rights, exp_summary_model,
+    skip_exploration_model_last_updated=False):
     """Create an ExplorationSummary domain object for a given Exploration
     domain object and return it.
 
     Args:
         exploration: Exploration. The exploration whose summary is to be
             computed.
+        exp_rights: ExplorationRightsModel. The exploration rights model, used
+            to compute summary.
+        exp_summary_model: ExplorationSummaryModel. The exploration summary
+            model, whose summary is computed.
+        skip_exploration_model_last_updated: bool. This flag is used to
+            update exploration_model_last_updated, only when required.
 
     Returns:
         ExplorationSummary. The resulting exploration summary domain object.
+
+    Raises:
+        Exception. ExplorationSummaryModel cannot be None and
+            skip_exploration_model_last_updated cannot be set to True
+            at the same time.
     """
-    exp_rights = exp_models.ExplorationRightsModel.get_by_id(exploration.id)
-    exp_summary_model = exp_models.ExpSummaryModel.get_by_id(exploration.id)
     if exp_summary_model:
         old_exp_summary = exp_fetchers.get_exploration_summary_from_model(
             exp_summary_model)
         ratings = old_exp_summary.ratings or feconf.get_empty_ratings()
         scaled_average_rating = get_scaled_average_rating(ratings)
+        exploration_model_last_updated = (
+            old_exp_summary.exploration_model_last_updated)
     else:
+        if skip_exploration_model_last_updated:
+            raise Exception(
```

Comment:
I think the docstring is wrong, it says this is an ExplorationSummaryModel when it should really be ExplorationSummaryModel|None.

Also I'm concerned about passing the models around between functions -- that's a pattern we try to avoid. Can we not do that? What was problematic about the old code (which retrieved the summary model within the same function)?

Reply:
> I think the docstring is wrong, it says this is an ExplorationSummaryModel when it should really be ExplorationSummaryModel|None.

Right.

> Also I'm concerned about passing the models around between functions -- that's a pattern we try to avoid. Can we not do that? What was problematic about the old code (which retrieved the summary model within the same function)?

We need this function or the job so that we can get the models using Beam IO and then pass them here, but I agree that the rights can be not-model.